### ISSUE TITLE
[1.8.1-beta.3] stale client data using reactive-publish [simple repro attached]

### ISSUE THREAD

**Issue by wildhart** [meteor/meteor#10447]
I'm using `peerlibrary:reactive-publish` to implement SAAS whereby data is only published to the client which matches a user's "group".

This was working fine in Meteor 1.8.0.2 but since testing 1.8.1-beta.16 I noticed that when switching group the data matching the old group is not removed from the client.

Here's a very simple repro which clearly demonstrates the issue: [bug.zip](https://github.com/meteor/meteor/files/2849619/bug.zip)

**Steps to reproduce:**
1. Unzip the attached repro and `meteor run`.
2. Log in with username "test" and password "test"
3. Click on the 'Switch group' button a couple of times.
4. You will see this:
    ![bad](https://user-images.githubusercontent.com/12505024/52542354-af292680-2e03-11e9-8803-c84609df5740.png)

Since the current group is "a" the "Group B" item **should not be in the list**.

**Expected behaviour**
1. Kill the app.
2.  `meteor update --release 1.8.0.2 && meteor`
3. Click the button a couple of times and see the desired result.

Clicking the button switches the user's `profile.group` between "a" and "b".  The "todos" publication only publishes items which have a matching "group" property.  The list of todos shown on the client should only ever match the user's current group as it does with Meteor 1.8.0.2:
    ![good](https://user-images.githubusercontent.com/12505024/52542411-3b3b4e00-2e04-11e9-846e-d82e62d9a886.png)

You can see in the server console that the publication is responding to the change in the user's group and is only publishing the correct item.  The new item is sent to the client and added to minimongo, but the old item is never removed.

I'm using the latest version of [`peerlibrary:reactive-publish`](https://github.com/peerlibrary/meteor-reactive-publish) which is maintained by @mitar, but that is not changing between upgrading Meteor from 1.8.0.2 to 1.8.1-beta.16, so I'm reporting the bug here instead.

**Comment by wildhart** [comment]
I've narrowed this down to a change between 1.8.1-beta.2 and 1.8.1-beta.3.

Package changes between those two releases:
```
autoupdate     upgraded from 1.5.1 to 1.6.0-beta181.3
ddp-server     upgraded from 2.2.0 to 2.3.0-beta181.3
reactive-dict  upgraded from 1.2.1 to 1.3.0-beta181.3
session        upgraded from 1.1.8 to 1.2.0-beta181.3
```

**Comment by mitar** [comment]
What is a diff for `ddp-server` between versions? If you link to it I could take a look if I notice anything.

**Comment by wildhart** [comment]
I'm not an expert, but I think the only changes are in https://github.com/meteor/meteor/commit/79ae1849f5846e207ed4dfa49831ba409252fd03#diff-7e48a19069e07781db2a50282a8bfdd7 by @nathan-muir

**Comment by wildhart** [comment]
I've been able to confirm that the bug is definitely in that commit.  I've switched back and forth between the version `packages/ddp-server/livedata_server.js` either side of that commit and the bug only occurs with the newer version (both are 2.2.0).

There are lots of changes in that commit and I'm not familiar with `Set()` or `Map()` so wouldn't know where to start isolating the bug further.  @nathan-muir, can you help?

**Comment by nathan-muir** [comment]
Yep, those changes ( from https://github.com/meteor/meteor/pull/10053 ) converted a lot of `Object` to the es6 classes `Set` [1] & `Map` [2]; mostly to fix performance issues that occur when removing keys from an object.

While the changes were merged into master quite some time ago; they have not been correctly version bumped -- so most people haven't actually got them in `ddp-server` as yet.

As for updating `peerlibrary:reactive-publish` - it would mostly involve changing `obj[key]` into `map.get(key)`.

[1] https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set
[2] https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

**Comment by mitar** [comment]
Huh huh. There might a whole series of my packages to then change here. :-(

I would have to look into how to make this in a way to support both versions.

**Comment by wildhart** [comment]
I've opened a bug in `peerlibrary:reactive-publish` so that I can be notified when that package gets updated.

Just my $0.02: looking at the comments on #10053 there may be some other 3rd party packages out there which my break due to this 'internal' change in `ddp-client`.  Maybe this should be mentioned in the release notes for 1.8.1?...

**Comment by etyp** [comment]
Yeah @wildhart this is causing all sorts of trouble for us in trying out the 1.8.1 beta. These changes also impact people using meteorhacks:meteorx (see https://github.com/meteorhacks/meteorx/issues/8) and meteorhacks:unblock.

@benjamn what's the best way for me to contribute to add warnings to History.md so others aren't opening new issues right after the 1.8.1 release?
